// @egjs/axes/outjs/const.js
export var DIRECTION_NONE = 1;
export var DIRECTION_LEFT = 2;
export var DIRECTION_RIGHT = 4;
export var DIRECTION_HORIZONTAL = 2 | 4;
export var DIRECTION_UP = 8;
export var DIRECTION_DOWN = 16;
export var DIRECTION_VERTICAL = 8 | 16;
export var DIRECTION_ALL = 2 | 4 | 8 | 16;
export var MOUSE_LEFT = "left";
export var MOUSE_RIGHT = "right";
export var MOUSE_MIDDLE = "middle";
export var ANY = "any";
export var NONE = "none";
export var SHIFT = "shift";
export var CTRL = "ctrl";
export var ALT = "alt";
export var META = "meta";
export var VELOCITY_INTERVAL = 16;
export var AXES_METHODS = [
    "connect",
    "disconnect",
    "get",
    "setTo",
    "setBy",
    "setOptions",
    "setAxis",
    "stopAnimation",
    "updateAnimation",
    "isBounceArea",
];
export var AXES_EVENTS = [
    "hold",
    "release",
    "change",
    "animationStart",
    "animationEnd",
    "finish",
];
import getAgent from "@egjs/agent";
import { window } from "./browser";
export var IOS_EDGE_THRESHOLD = 30;
export var IS_IOS_SAFARI = "ontouchstart" in window && getAgent().browser.name === "safari";
export var TRANSFORM = (function () {
    if (typeof document === "undefined") {
        return "";
    }
    var bodyStyle = (document.head || document.getElementsByTagName("head")[0])
        .style;
    var target = [
        "transform",
        "webkitTransform",
        "msTransform",
        "mozTransform",
    ];
    for (var i = 0, len = target.length; i < len; i++) {
        if (target[i] in bodyStyle) {
            return target[i];
        }
    }
    return "";
})();
export var PREVENT_DRAG_CSSPROPS = {
    "-webkit-user-select": "none",
    "-ms-user-select": "none",
    "-moz-user-select": "none",
    "user-select": "none",
    "-webkit-user-drag": "none",
};
